### PR TITLE
[monodroid] Introduce 'win32 host' BCL build

### PR DESCRIPTION
**Pull request by alexischr**
TODO:
- [x] Check tests
- [x] Set up CI for `make -C runtime build_profiles="monodroid" PROFILE_PLATFORM="win32"`  (on win32)

### CODE REVIEW

**Comment by alexischr** [comment]
Current issue: Whille `referencesource/System.Xml/System/Xml/Serialization/ImportContext.cs` is included in the `System.Xml.dll` response file, other files complain about missing the public class:

```
C:\src\mono\mcs\class\referencesource\System.Xml\System\Xml\Serialization\XmlSchemaImporter.cs(63,115): error CS0246: The type or namespace name 'ImportContext' could not be found (are you missing a using directive or an assembly reference?)
```

File: mcs/build/profiles/monodroid.make [hunk 1]
```diff
@@ -26,11 +26,17 @@ PROFILE_MCS_FLAGS = \
...
 # the tuner takes care of the install
 NO_INSTALL = yes
 MOBILE_DYNAMIC = yes
 MOBILE_PROFILE = yes
 NO_CONSOLE = yes
+
+ifdef PLATFORM_WIN32
+PROFILE_MCS_FLAGS += -d:MOBILE_DESKTOP_HOST -d:WIN_PLATFORM
+MOBILE_PROFILE=
+FRAMEWORK_VERSION= 4.5
```
Comment (marek-safar) [inline mcs/build/profiles/monodroid.make]: This is wrong it's still 2.1 version

File: mcs/build/profiles/monodroid.make [hunk 2]
```diff
@@ -26,11 +26,17 @@ PROFILE_MCS_FLAGS = \
 	-nostdlib \
 	$(PLATFORM_DEBUG_FLAGS)
 
-API_BIN_PROFILE = build/monodroid
```
Comment (marek-safar) [inline mcs/build/profiles/monodroid.make]: We cannot add reference to 4.0 assemblies

File: mcs/build/profiles/monodroid.make (hunk 1 above)
Comment (alexischr) [reply mcs/build/profiles/monodroid.make]: Yeah I have since removed that line

**Comment by lambdageek** [comment]
@alexischr `mono_fsw.sources` is missing from the PR.

```
MONO_PATH="./../../class/lib/build;$MONO_PATH" /cygdrive/c/mono-build/mono/sdks/builds/bcl/runtime/mono-wrapper  ./../../class/lib/build/gensources.exe --strict --platformsdir:./../../build "../../build/deps/win32_monodroid__System.dll.sources" "System.dll" "win32" "monodroid"
// Include does not exist: C:\mono-build\mono\mcs\class\System\mono_fsw.sources
// gensources produced 1 error(s) and a set of 1009 filename(s)
// Invoked with 'C:\mono-build\mono\mcs\class\lib\build\gensources.exe --strict --platformsdir:./../../build ../../build/deps/win32_monodroid__System.dll.sources System.dll win32 monodroid'
// Working directory was 'C:\mono-build\mono\mcs\class\System'
make[9]: *** No rule to make target '../../build/deps/win32_monodroid__System.dll.sources', needed by '../../build/deps/win32_monodroid__System.dll.response'.  Stop.
```

**Comment by lambdageek** [comment]
There's a merge issue with master since https://github.com/mono/mono/commit/aedaf15f278fc08c9b03321e939c8b3a4bb8e614 removed `mono_fsw.sources`:

```
MONO_PATH="./../../class/lib/build;$MONO_PATH" /cygdrive/c/mono-build/mono/sdks/builds/bcl/runtime/mono-wrapper  ./../../class/lib/build/gensources.exe --strict --platformsdir:./../../build "../../build/deps/win32_monodroid__System.dll.sources" "System.dll" "win32" "monodroid"
// Include does not exist: C:\mono-build\mono\mcs\class\System\mono_fsw.sources
// gensources produced 1 error(s) and a set of 1009 filename(s)
// Invoked with 'C:\mono-build\mono\mcs\class\lib\build\gensources.exe --strict --platformsdir:./../../build ../../build/deps/win32_monodroid__System.dll.sources System.dll win32 monodroid'
// Working directory was 'C:\mono-build\mono\mcs\class\System'
make[9]: *** No rule to make target '../../build/deps/win32_monodroid__System.dll.sources', needed by '../../build/deps/win32_monodroid__System.dll.response'.
```

**Comment by lambdageek** [comment]
@lexas please cherrypick https://github.com/lambdageek/mono/commit/04f42a3aaf36b17cfa75eaf5568f71b205685731 so we can try `@monojenkins build Windows x64 Android SDK` on this PR.

**Comment by alexischr** [comment]
@monojenkins build Windows x64 Android SDK

**Comment by lambdageek** [comment]
"Windows x64 Android SDK" is dying when trying to build the tests.
Because I forgot to pass the `TEST_PROFILE=win32` stuff in that part of the BclTemplate.

I'll push a fix.

```
CSC     [monodroid] mscorlib.dll
System\TimeZoneInfo.cs(153,23): error CS0111: Type 'TimeZoneInfo' already defines a member called 'CreateLocal' with the same parameter types
System\TimeZoneInfo.cs(197,23): error CS0111: Type 'TimeZoneInfo' already defines a member called 'FindSystemTimeZoneByIdCore' with the same parameter types
System\TimeZoneInfo.cs(207,15): error CS0111: Type 'TimeZoneInfo' already defines a member called 'GetSystemTimeZonesCore' with the same parameter types
make[8]: *** [../../build/library.make:335: ../../class/lib/monodroid/mscorlib.dll] Error 1
make[8]: Leaving directory '/cygdrive/d/j/workspace/xa/mcs/class/corlib'
make[7]: *** [../../build/rules.make:223: do-xunit-test] Error 2
make[7]: Leaving directory '/cygdrive/d/j/workspace/xa/mcs/class/corlib'
make[6]: *** [../build/rules.make:244: xunit-test-recursive] Error 1
make[6]: Leaving directory '/cygdrive/d/j/workspace/xa/mcs/class'
make[5]: *** [build/rules.make:244: xunit-test-recursive] Error 1
make[5]: Leaving directory '/cygdrive/d/j/workspace/xa/mcs'
make[4]: *** [Makefile:57: profile-do--monodroid--xunit-test] Error 2
make[4]: Leaving directory '/cygdrive/d/j/workspace/xa/mcs'
make[3]: *** [Makefile:53: profiles-do--xunit-test] Error 2
make[3]: Leaving directory '/cygdrive/d/j/workspace/xa/mcs'
```

**Comment by alexischr** [comment]
@lambdageek I noticed the same thing locally, that not passing PROFILE_PLATFORM when I was building would build "differently". I don't know why but the non `win32`-specialized monodroid profile seems to be getting built, but with the defines `WIN_PLATFORM` and `MOBILE_DESKTOP_HOST` still turned on. I would think the platform would default to `win32` in all ways and that setting this manually would not be necessary on Windows.

**Comment by lambdageek** [comment]
@monojenkins build Windows x64 Android SDK

**Comment by lambdageek** [comment]
Dying in `CSC     [monodroid-win32] monodroid_System_xunit-test.dll`
Looks like it's including tests from `monodroid_System_xtest.dll.exclude.sources`.  I tried making a `win32_monodroid_System_xtest.dll.exclude.sources` file but it didn't seem to help.

**Comment by alexischr** [comment]
@monojenkins build Windows x64 Android SDK

**Comment by alexischr** [comment]
@lambdageek success!! https://jenkins.mono-project.com/job/xa/18/Azure/processDownloadRequest/18/android-release-Windows-7e2fd2d07cf6ad8d497361010837f93b077759c2.zip contains the profile. Will do some touchup but feel free to approve :)

**Comment by alexischr** [comment]
@monojenkins build Windows x64 Android SDK

**Comment by alexischr** [comment]
For the sake of integration and the oddities of replacing a profile that was problematic and untested in the first place, this will land in two or more patches. As of this patch, there are still failing tests, but all assemblies build and pass most tests. 

A prototype profile-building/testing Jenkins job is here: https://jenkins.mono-project.com/job/x-monodroid/

**Comment by alexischr** [comment]
@monojenkins backport to 2019-04

**Comment by alexischr** [comment]
@monojenkins backport to 2019-02

**Comment by alexischr** [comment]
@monojenkins commit csproj

**Pull request merged**